File: levelingHandler.py
"""Upgrade-card presentation and input handling."""

import pygame

import upgrades
import uiTheme as ui
import variableHolster as vH


class LevelingHandler:
    CARD_KEYS = (pygame.K_1, pygame.K_2, pygame.K_3)

    def __init__(self):
        self.frameRate = vH.frameRate
        self.sW, self.sH = vH.screen.get_size()
        self.cardColor = ui.PANEL
        self.cardHoverColor = ui.PANEL_HOVER
        self.baseColor = ui.TEXT
        self.rerolls = 2
        self.tileSize = min(self.sW, self.sH) / 20
        self.uiScale = ui.display_scale(vH.screen)
        font_path = "data/media/coolveticarg.otf"
        self.titleFont = pygame.font.Font(font_path, int(self.tileSize * 0.9))
        self.descFont = pygame.font.Font(font_path, int(self.tileSize * 0.45))
        self.smallFont = pygame.font.Font(font_path, int(self.tileSize * 0.34))
        self.update_layout()
        self.firstClick = True
        self.upgradeRarityColors = ui.RARITY_COLORS
        self.upgradeRarity = dict(upgrades.RARITY_MULTIPLIERS)
        self.upgradeTypesList = list(upgrades.DEFINITIONS_BY_NAME)
        self.upgradeBasicTypesAdd = {
            item.name: item.additive for item in upgrades.DEFINITIONS
        }
        self.upgradeBasicTypesMult = {
            item.name: item.multiplicative for item in upgrades.DEFINITIONS
        }
        self.cards = upgrades.generate_offer(count=3)
        self.selected_card = None
        self.randomizing = False
        self._sync_legacy_fields()

    def _px(self, value):
        return max(1, int(round(value * self.uiScale)))

    def update_layout(self):
        card_width = (self.sW - self.tileSize * 5) / 3
        card_height = self.sH * 0.62
        card_y = (self.sH - card_height) / 2 - self.tileSize * 0.35
        self.rerollButton = pygame.Rect(
            self.sW // 2 - self.tileSize * 2.1,
            self.sH * 0.87,
            self.tileSize * 4.2,
            self.tileSize,
        )
        self.leftCard = pygame.Rect(self.tileSize * 2, card_y, card_width, card_height)
        self.midCard = pygame.Rect(self.tileSize * 3 + card_width, card_y, card_width, card_height)
        self.rightCard = pygame.Rect(self.tileSize * 4 + 2 * card_width, card_y, card_width, card_height)
        self.card_rects = (self.leftCard, self.midCard, self.rightCard)

    def _sync_legacy_fields(self):
        """Keep the original public fields available while old callers migrate."""
        sides = ("left", "mid", "right")
        for side, card in zip(sides, self.cards):
            setattr(self, f"{side}CardUpgradeRarity", card.rarity)
            setattr(self, f"{side}CardUpgradeMath", card.math_type)
            setattr(self, f"{side}CardUpgradeType", card.name)

    def _draw_centered(self, font, text, color, x, y):
        rendered = font.render(text, True, color)
        vH.screen.blit(rendered, rendered.get_rect(center=(x, y)))

    def _projected_value(self, card, cS):
        base = cS.collectiveStats[card.name]
        additive = sum(cS.collectiveAddStats[card.name])
        multiplicative = 1
        for value in cS.collectiveMultStats[card.name]:
            multiplicative *= value
        modifier = upgrades.card_modifier(card)
        current = (base + additive) * multiplicative
        projected = ((base + additive + modifier) * multiplicative
                     if card.math_type == "additive"
                     else current * modifier)
        return current, projected

    def _recommendation(self, card, cS):
        owned_categories = {}
        for name, count in cS.upgradeCollection["types"].items():
            definition = upgrades.DEFINITIONS_BY_NAME.get(name)
            if definition:
                owned_categories[definition.category] = owned_categories.get(definition.category, 0) + count
        if cS.healthPoints <= cS.maxHealthPoints * .45 and card.definition.category == "survival":
            return "SAFE PICK", ui.GREEN
        if owned_categories.get(card.definition.category, 0) >= 2:
            return "BUILD MATCH", ui.PURPLE
        return None, None

    def drawCards(self):
        import characterStats as cS

        vH.screen.fill(ui.VOID)
        grid_size = max(24, int(self.tileSize * 0.55))
        for x in range(0, int(self.sW), grid_size):
            pygame.draw.line(vH.screen, pygame.Color(23, 27, 35), (x, 0), (x, self.sH))
        for y in range(0, int(self.sH), grid_size):
            pygame.draw.line(vH.screen, pygame.Color(23, 27, 35), (0, y), (self.sW, y))

        ui.draw_text(vH.screen, "LEVEL SECURED", self.tileSize * 0.34, ui.GREEN, (self.sW / 2, self.tileSize * 0.3), "midtop")
        ui.draw_text(vH.screen, "CHOOSE ONE // SHAPE THE RUN", self.tileSize * 0.72, ui.TEXT, (self.sW / 2, self.tileSize * 0.75), "midtop")
        ui.draw_text(vH.screen, "Every card stays with you until the run ends.", self.tileSize * 0.3, ui.MUTED, (self.sW / 2, self.tileSize * 1.65), "midtop")
        if cS.pendingLevelUps > 1:
            ui.draw_tag(vH.screen, f"{cS.pendingLevelUps} DRAFTS QUEUED",
                        (self.tileSize * 2, self.tileSize * 1.25), ui.GOLD, int(self.tileSize * .21))

        mouse_position = (vH.mouseX, vH.mouseY)
        for index, (rect, card) in enumerate(zip(self.card_rects, self.cards)):
            hovered = rect.collidepoint(mouse_position)
            accent = self.upgradeRarityColors[card.rarity]
            pressed = hovered and vH.mouseDown
            visual_rect = rect.move(0, self._px(2) if pressed else (-self._px(7) if hovered else 0))
            ui.draw_panel(vH.screen, visual_rect, self.cardColor, accent if hovered else ui.BORDER, shadow=3 if pressed else 7, hovered=hovered)
            pygame.draw.rect(vH.screen, accent, (visual_rect.x, visual_rect.y, visual_rect.width, self._px(9)))
            key_rect = pygame.Rect(visual_rect.x + self._px(18), visual_rect.y + self._px(24), self._px(38), self._px(38))
            pygame.draw.rect(vH.screen, ui.INK, key_rect)
            pygame.draw.rect(vH.screen, accent, key_rect, self._px(2))
            ui.draw_text(vH.screen, str(index + 1), self.tileSize * 0.4, accent, key_rect.center, "center")
            rarity_width = ui.font(int(self.tileSize * .23)).size(card.rarity.upper())[0]
            ui.draw_tag(vH.screen, card.rarity, (visual_rect.right - self._px(22) - rarity_width, visual_rect.y + self._px(29)), accent, int(self.tileSize * .23))
            ui.draw_text(vH.screen, card.definition.category.upper() + " CARD", self.tileSize * 0.24, ui.MUTED, (visual_rect.centerx, visual_rect.y + self.tileSize * 1.55), "center")
            ui.draw_text(vH.screen, card.name, self.tileSize * 0.58, ui.TEXT, (visual_rect.centerx, visual_rect.y + self.tileSize * 2.15), "center")
            pygame.draw.line(vH.screen, accent, (visual_rect.x + self._px(28), visual_rect.y + self.tileSize * 2.72), (visual_rect.right - self._px(28), visual_rect.y + self.tileSize * 2.72), self._px(2))
            ui.draw_text(vH.screen, card.definition.description, self.tileSize * 0.38, self.baseColor, (visual_rect.centerx, visual_rect.centery - self.tileSize * 0.2), "center")
            mode = "Flat bonus" if card.math_type == "additive" else "Scaling bonus"
            mode_width = ui.font(int(self.tileSize * .2)).size(mode.upper())[0]
            ui.draw_tag(vH.screen, mode, (visual_rect.centerx - mode_width / 2, visual_rect.centery + self.tileSize * 0.65), ui.BLUE, int(self.tileSize * .2))
            ui.draw_text(vH.screen, upgrades.format_card_value(card), self.tileSize * 0.78, accent, (visual_rect.centerx, visual_rect.bottom - self.tileSize * 1.25), "center")
            current, projected = self._projected_value(card, cS)
            direction = "LOWER IS FASTER" if card.name == "Attack Speed" else "PROJECTED STAT"
            ui.draw_text(vH.screen, f"{direction}  //  {current:.2f} → {projected:.2f}",
                         self.tileSize * .19, ui.GREEN if projected != current else ui.MUTED,
                         (visual_rect.centerx, visual_rect.bottom - self.tileSize * .76), "center")
            owned = cS.upgradeCollection["types"].get(card.name, 0)
            ui.draw_text(vH.screen, f"OWNED  {owned}", self.tileSize * 0.22, ui.MUTED, (visual_rect.centerx, visual_rect.bottom - self.tileSize * 0.45), "center")
            recommendation, recommendation_color = self._recommendation(card, cS)
            if recommendation:
                ui.draw_tag(vH.screen, recommendation,
                            (visual_rect.x + self._px(18), visual_rect.bottom - self.tileSize * .62),
                            recommendation_color, int(self.tileSize * .18))

        ui.draw_button(
            vH.screen, self.rerollButton, f"RUN REROLLS  //  {self.rerolls} LEFT", mouse_position,
            vH.mouseDown, self.rerolls > 0, ui.RED, "R", int(self.tileSize * 0.31),
        )

    def PlayerClicked(self):
        if pygame.K_r in vH.keyPressed and self.rerolls > 0:
            self.randomizeLevelUp()
            self.rerolls -= 1
            return "none"

        for index, key in enumerate(self.CARD_KEYS):
            if key in vH.keyPressed:
                self.selected_card = self.cards[index]
                return ("leftCard", "midCard", "rightCard")[index]

        if not vH.mouseDown:
            self.firstClick = False
        if vH.mouseDown and not self.firstClick:
            if self.rerollButton.collidepoint(vH.mouseX, vH.mouseY):
                self.firstClick = True
                if self.rerolls > 0:
                    self.randomizeLevelUp()
                    self.rerolls -= 1
                return "none"
            for index, rect in enumerate(self.card_rects):
                clickable_rect = rect.inflate(0, self._px(14)).move(0, -self._px(7))
                if clickable_rect.collidepoint(vH.mouseX, vH.mouseY):
                    self.firstClick = True
                    self.selected_card = self.cards[index]
                    return ("leftCard", "midCard", "rightCard")[index]
        return "none"

    def randomizeLevelUp(self):
        # Import locally to avoid a module cycle during characterStats initialization.
        import characterStats as cS

        self.cards = upgrades.generate_offer(cS.upgradeCollection, count=3)
        self.selected_card = None
        self._sync_legacy_fields()
        self.randomizing = True

File: RotBoiRemastered/UI/LevelingHandler.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Input;
using RotBoiRemastered.Core;
using RotBoiRemastered.Systems;

namespace RotBoiRemastered.UI;

/// <summary>
/// Everything LevelingHandler's stat-preview and recommendation logic needs
/// from the run's build state. Ported from the `cS` (characterStats.py)
/// reads inside levelingHandler.py's `_projected_value`/`_recommendation`/
/// `drawCards` -- characterStats.py itself isn't ported yet (see
/// Entities/README.md's Player.cs note), so this snapshot is the explicit
/// seam: whatever eventually owns real run state builds one of these each
/// frame instead of LevelingHandler reaching into a global.
/// </summary>
public sealed class LevelUpStatSnapshot
{
    public required IReadOnlyDictionary<string, double> CollectiveStats { get; init; }
    public required IReadOnlyDictionary<string, IReadOnlyList<double>> CollectiveAddStats { get; init; }
    public required IReadOnlyDictionary<string, IReadOnlyList<double>> CollectiveMultStats { get; init; }
    public required IReadOnlyDictionary<string, int> UpgradeTypeCounts { get; init; }
    public double HealthPoints { get; init; }
    public double MaxHealthPoints { get; init; }
    public int PendingLevelUps { get; init; }
}

/// <summary>
/// Upgrade-card presentation and input handling. Ported from
/// levelingHandler.py.
///
/// Cleanup vs. the Python original: dropped several fields that were
/// assigned in `__init__` but never read anywhere else in the file or the
/// rest of the repo (verified by grep) -- `titleFont`/`descFont`/
/// `smallFont` (every draw call resolves its font fresh through
/// `UiTheme.Font` instead), `cardHoverColor` (`DrawPanel`'s own `hovered`
/// flag already picks `PanelHover` internally), `upgradeRarity`/
/// `upgradeTypesList`/`upgradeBasicTypesAdd`/`upgradeBasicTypesMult`, and
/// `frameRate`. Also dropped `_draw_centered` (defined, never called) and
/// `_sync_legacy_fields` (a Python-side "keep old attribute names working
/// while callers migrate" shim with nothing left in this codebase to
/// migrate).
/// </summary>
public sealed class LevelingHandler
{
    private static readonly Keys[] CardKeys = { Keys.D1, Keys.D2, Keys.D3 };
    private static readonly string[] CardSlotNames = { "leftCard", "midCard", "rightCard" };

    public int Rerolls { get; private set; } = 2;
    public List<UpgradeCard> Cards { get; private set; }
    public UpgradeCard? SelectedCard { get; private set; }
    public bool Randomizing { get; private set; }

    public Rectangle RerollButton { get; private set; }
    public Rectangle LeftCard { get; private set; }
    public Rectangle MidCard { get; private set; }
    public Rectangle RightCard { get; private set; }
    public IReadOnlyList<Rectangle> CardRects { get; private set; } = Array.Empty<Rectangle>();

    private int _screenWidth, _screenHeight;
    private float _tileSize;
    private bool _firstClick = true;

    public LevelingHandler(int screenWidth, int screenHeight, Random? rng = null)
    {
        Cards = Upgrades.GenerateOffer(count: 3, rng: rng);
        UpdateLayout(screenWidth, screenHeight);
    }

    private float Px(float value) => Math.Max(1, MathF.Round(value * UiTheme.DisplayScale(_screenWidth, _screenHeight)));

    public void UpdateLayout(int screenWidth, int screenHeight)
    {
        _screenWidth = screenWidth;
        _screenHeight = screenHeight;
        _tileSize = Math.Min(screenWidth, screenHeight) / 20f;

        float cardWidth = (screenWidth - _tileSize * 5) / 3f;
        float cardHeight = screenHeight * 0.62f;
        float cardY = (screenHeight - cardHeight) / 2f - _tileSize * 0.35f;

        RerollButton = new Rectangle(
            (int)(screenWidth / 2f - _tileSize * 2.1f), (int)(screenHeight * 0.87f),
            (int)(_tileSize * 4.2f), (int)_tileSize);
        LeftCard = new Rectangle((int)(_tileSize * 2), (int)cardY, (int)cardWidth, (int)cardHeight);
        MidCard = new Rectangle((int)(_tileSize * 3 + cardWidth), (int)cardY, (int)cardWidth, (int)cardHeight);
        RightCard = new Rectangle((int)(_tileSize * 4 + 2 * cardWidth), (int)cardY, (int)cardWidth, (int)cardHeight);
        CardRects = new[] { LeftCard, MidCard, RightCard };
    }

    /// <summary>Public (Python's `_projected_value` was underscore-private in name only).</summary>
    public static (double Current, double Projected) ProjectedValue(UpgradeCard card, LevelUpStatSnapshot stats)
    {
        double baseValue = stats.CollectiveStats[card.Name];
        double additive = stats.CollectiveAddStats[card.Name].Sum();
        double multiplicative = stats.CollectiveMultStats[card.Name].Aggregate(1.0, (acc, v) => acc * v);
        double modifier = Upgrades.CardModifier(card);
        double current = (baseValue + additive) * multiplicative;
        double projected = card.MathType == "additive"
            ? (baseValue + additive + modifier) * multiplicative
            : current * modifier;
        return (current, projected);
    }

    /// <summary>Public (Python's `_recommendation` was underscore-private in name only).</summary>
    public static (string? Label, Color? Accent) Recommendation(UpgradeCard card, LevelUpStatSnapshot stats)
    {
        var ownedCategories = new Dictionary<string, int>();
        foreach (var (name, count) in stats.UpgradeTypeCounts)
        {
            if (Upgrades.DefinitionsByName.TryGetValue(name, out var definition))
                ownedCategories[definition.Category] = ownedCategories.GetValueOrDefault(definition.Category) + count;
        }
        if (stats.HealthPoints <= stats.MaxHealthPoints * .45 && card.Definition.Category == "survival")
            return ("SAFE PICK", UiTheme.Green);
        if (ownedCategories.GetValueOrDefault(card.Definition.Category) >= 2)
            return ("BUILD MATCH", UiTheme.Purple);
        return (null, null);
    }

    public void DrawCards(SpriteBatch spriteBatch, LevelUpStatSnapshot stats, Point mousePosition, bool mouseDown)
    {
        Primitives2D.FillRect(spriteBatch, new Rectangle(0, 0, _screenWidth, _screenHeight), UiTheme.Void);
        int gridSize = Math.Max(24, (int)(_tileSize * 0.55f));
        var gridColor = new Color(23, 27, 35);
        for (int x = 0; x < _screenWidth; x += gridSize)
            Primitives2D.Line(spriteBatch, new Vector2(x, 0), new Vector2(x, _screenHeight), gridColor, 1);
        for (int y = 0; y < _screenHeight; y += gridSize)
            Primitives2D.Line(spriteBatch, new Vector2(0, y), new Vector2(_screenWidth, y), gridColor, 1);

        UiTheme.DrawText(spriteBatch, "LEVEL SECURED", _tileSize * 0.34, UiTheme.Green,
            new Vector2(_screenWidth / 2f, _tileSize * 0.3f), "midtop");
        UiTheme.DrawText(spriteBatch, "CHOOSE ONE // SHAPE THE RUN", _tileSize * 0.72, UiTheme.Text,
            new Vector2(_screenWidth / 2f, _tileSize * 0.75f), "midtop");
        UiTheme.DrawText(spriteBatch, "Every card stays with you until the run ends.", _tileSize * 0.3, UiTheme.Muted,
            new Vector2(_screenWidth / 2f, _tileSize * 1.65f), "midtop");
        if (stats.PendingLevelUps > 1)
        {
            UiTheme.DrawTag(spriteBatch, $"{stats.PendingLevelUps} DRAFTS QUEUED",
                new Vector2(_tileSize * 2, _tileSize * 1.25f), UiTheme.Gold, _tileSize * .21);
        }

        double textScale = UiTheme.TextScaleMultiplier();
        for (int index = 0; index < CardRects.Count; index++)
        {
            var rect = CardRects[index];
            var card = Cards[index];
            bool hovered = rect.Contains(mousePosition);
            Color accent = UiTheme.RarityColors.TryGetValue(card.Rarity, out var rarityColor) ? rarityColor : UiTheme.Border;
            bool pressed = hovered && mouseDown;
            var visualRect = new Rectangle(rect.X, rect.Y + (int)(pressed ? Px(2) : hovered ? -Px(7) : 0), rect.Width, rect.Height);
            UiTheme.DrawPanel(spriteBatch, visualRect, UiTheme.Panel, hovered ? accent : UiTheme.Border,
                shadow: pressed ? 3 : 7, hovered: hovered);
            Primitives2D.FillRect(spriteBatch, new Rectangle(visualRect.X, visualRect.Y, visualRect.Width, (int)Px(9)), accent);

            var keyRect = new Rectangle((int)(visualRect.X + Px(18)), (int)(visualRect.Y + Px(24)), (int)Px(38), (int)Px(38));
            Primitives2D.FillRect(spriteBatch, keyRect, UiTheme.Ink);
            Primitives2D.RectOutline(spriteBatch, keyRect, accent, (int)Px(2));
            UiTheme.DrawText(spriteBatch, index + 1, _tileSize * 0.4, accent,
                new Vector2(keyRect.Center.X, keyRect.Center.Y), "center");

            float rarityWidth = UiTheme.Font(_tileSize * .23).MeasureString(card.Rarity.ToUpperInvariant()).X;
            UiTheme.DrawTag(spriteBatch, card.Rarity,
                new Vector2(visualRect.Right - Px(22) - rarityWidth, visualRect.Y + Px(29)), accent, _tileSize * .23);
            UiTheme.DrawText(spriteBatch, card.Definition.Category.ToUpperInvariant() + " CARD", _tileSize * 0.24, UiTheme.Muted,
                new Vector2(visualRect.Center.X, visualRect.Y + _tileSize * 1.55f), "center");
            UiTheme.DrawText(spriteBatch, card.Name, _tileSize * 0.58, UiTheme.Text,
                new Vector2(visualRect.Center.X, visualRect.Y + _tileSize * 2.15f), "center");
            Primitives2D.Line(spriteBatch, new Vector2(visualRect.X + Px(28), visualRect.Y + _tileSize * 2.72f),
                new Vector2(visualRect.Right - Px(28), visualRect.Y + _tileSize * 2.72f), accent, (int)Px(2));
            UiTheme.DrawText(spriteBatch, card.Definition.Description, _tileSize * 0.38, UiTheme.Text,
                new Vector2(visualRect.Center.X, visualRect.Center.Y - _tileSize * 0.2f), "center");

            string mode = card.MathType == "additive" ? "Flat bonus" : "Scaling bonus";
            float modeWidth = UiTheme.Font(_tileSize * .2).MeasureString(mode.ToUpperInvariant()).X;
            UiTheme.DrawTag(spriteBatch, mode,
                new Vector2(visualRect.Center.X - modeWidth / 2, visualRect.Center.Y + _tileSize * 0.65f), UiTheme.Blue, _tileSize * .2);
            UiTheme.DrawText(spriteBatch, Upgrades.FormatCardValue(card), _tileSize * 0.78, accent,
                new Vector2(visualRect.Center.X, visualRect.Bottom - _tileSize * 1.25f * (float)textScale), "center");

            var (current, projected) = ProjectedValue(card, stats);
            string direction = card.Name == "Attack Speed" ? "LOWER IS FASTER" : "PROJECTED STAT";
            UiTheme.DrawText(spriteBatch, $"{direction}  //  {current:F2} → {projected:F2}", _tileSize * .19,
                projected != current ? UiTheme.Green : UiTheme.Muted,
                new Vector2(visualRect.Center.X, visualRect.Bottom - _tileSize * .76f * (float)textScale), "center");

            int owned = stats.UpgradeTypeCounts.GetValueOrDefault(card.Name);
            UiTheme.DrawText(spriteBatch, $"OWNED  {owned}", _tileSize * 0.22, UiTheme.Muted,
                new Vector2(visualRect.Center.X, visualRect.Bottom - _tileSize * 0.45f * (float)textScale), "center");

            var (recommendation, recommendationColor) = Recommendation(card, stats);
            if (recommendation is not null)
            {
                UiTheme.DrawTag(spriteBatch, recommendation,
                    new Vector2(visualRect.X + Px(18), visualRect.Bottom - _tileSize * .62f), recommendationColor, _tileSize * .18);
            }
        }

        UiTheme.DrawButton(spriteBatch, RerollButton, $"RUN REROLLS  //  {Rerolls} LEFT", mousePosition,
            mouseDown, Rerolls > 0, UiTheme.Red, "R", _tileSize * 0.31);
    }

    /// <summary>
    /// Returns "none" or one of "leftCard"/"midCard"/"rightCard" -- kept as
    /// the exact string contract levelingHandler.py's caller (character.py's
    /// not-yet-ported handleLevelingProcess) expects, since that's the only
    /// consumer and its final shape isn't decided yet.
    /// </summary>
    public string PlayerClicked(IReadOnlySet<Keys> keysPressed, Point mousePosition, bool mouseDown,
        IReadOnlyDictionary<string, int> upgradeTypeCounts, Random? rng = null)
    {
        if (keysPressed.Contains(Keys.R) && Rerolls > 0)
        {
            RandomizeLevelUp(upgradeTypeCounts, rng);
            Rerolls -= 1;
            return "none";
        }

        for (int index = 0; index < CardKeys.Length; index++)
        {
            if (keysPressed.Contains(CardKeys[index]))
            {
                SelectedCard = Cards[index];
                return CardSlotNames[index];
            }
        }

        if (!mouseDown)
            _firstClick = false;
        if (mouseDown && !_firstClick)
        {
            if (RerollButton.Contains(mousePosition))
            {
                _firstClick = true;
                if (Rerolls > 0)
                {
                    RandomizeLevelUp(upgradeTypeCounts, rng);
                    Rerolls -= 1;
                }
                return "none";
            }
            for (int index = 0; index < CardRects.Count; index++)
            {
                var clickable = CardRects[index];
                clickable.Inflate(0, (int)Px(14));
                clickable.Y -= (int)Px(7);
                if (clickable.Contains(mousePosition))
                {
                    _firstClick = true;
                    SelectedCard = Cards[index];
                    return CardSlotNames[index];
                }
            }
        }
        return "none";
    }

    public void RandomizeLevelUp(IReadOnlyDictionary<string, int> upgradeTypeCounts, Random? rng = null)
    {
        Cards = Upgrades.GenerateOffer(upgradeTypeCounts, count: 3, rng: rng);
        SelectedCard = null;
        Randomizing = true;
    }
}
